### PR TITLE
EOS-12131: SSPL: JBOD: Deployment: Control the list of resources to monitor (other monitoring/HA turned off)

### PR DESCRIPTION
**Pull request by pranav-seagate**
…

## Problem Statement
<pre>
  <code>
    Story Ref (if any):
    Please add Problem statement here...
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Please add Problem decription here...
  </code>
</pre>
## Solution
<pre>
  <code>
    Please add summary of the change,List any dependencies that are required for this change.
  </code>
</pre>
## Sanity testing on RPM done
<pre>
  <code>
    - [ ] Yes
    - [ ] No
  </code>
</pre>
## Unit/Manual Testing Description
<pre>
  <code>
    Please describe the tests that you ran to verify your changes.
  </code>
</pre>

### CODE REVIEW

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 [hunk 1]
```diff
@@ -101,6 +101,7 @@ IP_addr=172.16.1.222
 max_login_attempts=2
 
 [RAIDSENSOR]
+should_monitor=XXXXXX # RAIDSENSOR
```
Comment (indrajitzagade) [inline low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: Why "XXXXXX"? can we put true by default?

Comment (indrajitzagade) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: Same comment for others below.

File: low-level/framework/base/sspl_constants.py [hunk 2]
```diff
@@ -57,18 +57,52 @@
...
-            "IEMSensor", "NodeHWsensor",
-            "SASPortSensor", "MemFaultSensor", "CPUFaultSensor", "RAIDIntegritySensor"]
+        "REALSTORSENSORS": {
+            "ACTUATORS": [],
+            "CORE_PROCESSORS": [],
+            "DEGRADED_STATE_MODULES": [],
+            "MESSAGE_HANDLERS": [],
+            "SENSORS": ["RealStorFanSensor", "RealStorPSUSensor",
+                "RealStorControllerSensor", "RealStorDiskSensor", "RealStorSideplaneExpanderSensor",
+                "RealStorLogicalVolumeSensor", "RealStorEnclosureSensor",],
```
Comment (indrajitzagade) [inline low-level/framework/base/sspl_constants.py:67]: Remove comma "RealStorEnclosureSensor",

File: low-level/framework/base/sspl_constants.py [hunk 3]
```diff
@@ -57,18 +57,52 @@
...
+        },
+        "NODEHWSENSOR": {
+            "ACTUATORS": [],
+            "CORE_PROCESSORS": [],
+            "DEGRADED_STATE_MODULES": [ "NodeHWsensor",],
```
Comment (indrajitzagade) [inline low-level/framework/base/sspl_constants.py:72]: Remove comma "NodeHWsensor",

File: low-level/framework/base/sspl_constants.py [hunk 4]
```diff
@@ -57,18 +57,52 @@
...
+            "MESSAGE_HANDLERS": [],
+            "SENSORS": [ "NodeHWsensor",],
```
Comment (indrajitzagade) [inline low-level/framework/base/sspl_constants.py:74]: Same comment as line #72 above.

File: low-level/framework/base/sspl_constants.py [hunk 5]
```diff
@@ -57,18 +57,52 @@
...
+            "ACTUATORS": [],
+            "CORE_PROCESSORS": [],
+            "DEGRADED_STATE_MODULES": [],
+            "MESSAGE_HANDLERS": [],
+            "SENSORS": [],
+        },
+        "RAIDSENSOR": {
+            "ACTUATORS": [],
+            "CORE_PROCESSORS": [],
+            "DEGRADED_STATE_MODULES": ["RAIDsensor", ],
```
Comment (indrajitzagade) [inline low-level/framework/base/sspl_constants.py:86]: Same comment as line #72 above.

File: low-level/framework/base/sspl_constants.py [hunk 6]
```diff
@@ -57,18 +57,52 @@
...
+        "CATCHALL": {
+            "ACTUATORS" : ["Service", "RAIDactuator", "Smartctl", "NodeHWactuator", "RealStorActuator"],
+            "CORE_PROCESSORS" : ("RabbitMQegressProcessor", "RabbitMQingressProcessor", "LoggingProcessor"),
+            "DEGRADED_STATE_MODULES" : ("ServiceWatchdog", "NodeData", "IEMSensor",
+                "DiskMsgHandler", "LoggingMsgHandler", "ServiceMsgHandler", "NodeDataMsgHandler",
+                "NodeControllerMsgHandler", "SASPortSensor", "MemFaultSensor", "CPUFaultSensor"),
+            "MESSAGE_HANDLERS" : ("DiskMsgHandler", "LoggingMsgHandler", "ServiceMsgHandler", "NodeDataMsgHandler",
+                "NodeControllerMsgHandler", "RealStorEnclMsgHandler", "RealStorActuatorMsgHandler"),
+            "SENSORS" : ["ServiceWatchdog", "NodeData",  "IEMSensor",
+                "SASPortSensor", "MemFaultSensor", "CPUFaultSensor", ]
```
Comment (indrajitzagade) [inline low-level/framework/base/sspl_constants.py:99]: Same comment as line #72 above.

File: low-level/framework/base/sspl_constants.py [hunk 7]
```diff
@@ -57,18 +57,52 @@
...
+            "SENSORS": [],
+        },
+        "RAIDSENSOR": {
+            "ACTUATORS": [],
+            "CORE_PROCESSORS": [],
+            "DEGRADED_STATE_MODULES": ["RAIDsensor", ],
+            "MESSAGE_HANDLERS": [],
+            "SENSORS": ["RAIDsensor", "RAIDIntegritySensor"],
+        },
+        "CATCHALL": {
```
Comment (indrajitzagade) [inline low-level/framework/base/sspl_constants.py:90]: @pranav-seagate Will we disable e.g. SASportsensor, Mem/CPU fault sensors, etc. too for the virtual server?

File: low-level/framework/base/sspl_constants.py (hunk 2 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py]: Fixed.

File: low-level/framework/base/sspl_constants.py (hunk 3 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py]: Fixed

File: low-level/framework/base/sspl_constants.py (hunk 4 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py]: Fixed

File: low-level/framework/base/sspl_constants.py (hunk 5 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py]: Fixed

File: low-level/framework/base/sspl_constants.py (hunk 6 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py]: Fixed

File: low-level/framework/base/sspl_constants.py (hunk 7 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py:90]: At a quick glance these appear to be the normal threaded sensors, so it should be simple to add them to the list of configurable sensors by putting the 'should_monitor' key in the appropriate conf file section.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 1 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: In progress.

File: low-level/framework/base/sspl_constants.py (hunk 7 above)
Comment (indrajitzagade) [reply low-level/framework/base/sspl_constants.py:90]: We need to disable e.g. SASportsensor, Mem/CPU fault sensors, etc. too for the virtual server case, please make necessary changes if it is not done currently.

Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py:112]: Pushed.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 1 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1]: Fixed

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 [hunk 8]
```diff
@@ -101,6 +101,7 @@ IP_addr=172.16.1.222
 max_login_attempts=2
 
 [RAIDSENSOR]
+should_monitor=true # RAIDSENSOR
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: avoid comment for sensor name, its under sensor header already.

same comment elsewhere

File: low-level/files/opt/seagate/sspl/bin/conf_sensors_enable [hunk 9]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/conf_sensors_enable:1]: can we rename to conf_based_sensor_enable

File: low-level/files/opt/seagate/sspl/bin/conf_sensors_enable [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    [RAIDSENSOR]="true"
+    [SASPORTSENSOR]="true"
+    [MEMFAULTSENSOR]="true"
+    [CPUFAULTSENSOR]="true"
+
+)
+
+case "${storage_type}" in
+    virtual) 
+        should_monitor[REALSTORSENSORS]="false"
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/conf_sensors_enable:25]: should_monitor[] -> monitor[] can we keep, more simple

File: low-level/framework/base/sspl_constants.py [hunk 11]
```diff
@@ -57,18 +57,73 @@
...
-        "CORE_PROCESSORS" : ("RabbitMQegressProcessor", "RabbitMQingressProcessor", "LoggingProcessor"),
-        "DEGRADED_STATE_MODULES" : ("ServiceWatchdog", "RAIDsensor", "NodeData", "IEMSensor", "NodeHWsensor",
-                            "DiskMsgHandler", "LoggingMsgHandler", "ServiceMsgHandler", "NodeDataMsgHandler",
-                            "NodeControllerMsgHandler", "SASPortSensor", "MemFaultSensor", "CPUFaultSensor"),
-        "MESSAGE_HANDLERS" : ("DiskMsgHandler", "LoggingMsgHandler", "ServiceMsgHandler", "NodeDataMsgHandler",
-                        "NodeControllerMsgHandler"),
-        "SENSORS" : ["ServiceWatchdog", "RAIDsensor", "NodeData",
-            "IEMSensor", "NodeHWsensor",
-            "SASPortSensor", "MemFaultSensor", "CPUFaultSensor", "RAIDIntegritySensor"]
+        "REALSTORSENSORS": {
```
Comment (cdeshmukh) [inline low-level/framework/base/sspl_constants.py:60]: Had given this comment earlier also. 

Having actuators & handles under SENSORS category again doesnt make sense. Why cant we keep something generic, under which its ok to host sensors, actuators, handlers & core processors. One suggestion was 'REALSTORMONITOR'.

Same comment applies elsewhere.

File: low-level/framework/base/sspl_constants.py [hunk 12]
```diff
@@ -57,18 +57,73 @@
...
+        },
+        "CPUFAULTSENSOR": {
+            "ACTUATORS": [],
+            "CORE_PROCESSORS": [],
+            "DEGRADED_STATE_MODULES": ["CPUFaultSensor"],
+            "MESSAGE_HANDLERS": [],
+            "SENSORS": ["CPUFaultSensor"],
+        },
+
+        "CATCHALL": {
```
Comment (cdeshmukh) [inline low-level/framework/base/sspl_constants.py:112]: CATCHALL means this set will be enabled for all config types right ? ENABLE_ALWAYS name would suit ?

File: low-level/framework/sspl_ll_d [hunk 13]
```diff
@@ -414,11 +430,15 @@ def _sensors_actuators_factory(sspl_threaded_modules, msgQlist, operating_system
     """Loops thru list of sensors/actuators and instantiate"""
 
     # Read in list of sensors
-    sensors = SSPL_SETTINGS.get("SENSORS")
+    sensors = []
+    for group in sspl_settings_configured_groups:
+        sensors.extend(SSPL_SETTINGS[group].get("SENSORS"))
```
Comment (cdeshmukh) [inline low-level/framework/sspl_ll_d:435]: if no where printing final list, can we print here with INFO log level. Same for other module types being enabled finally

File: sspl_test/plans/sanity.pln [hunk 14]
```diff
@@ -1,4 +1,15 @@
 alerts.node.test_node_disk_actuator
+alerts.realstor.test_real_stor_controller_sensor
```
Comment (cdeshmukh) [inline sspl_test/plans/sanity.pln:2]: Good to see these back in sanity, we need this badly to keep out sanity really sane!

**Review by cdeshmukh — CHANGES_REQUESTED** [review]

File: low-level/framework/base/sspl_constants.py (hunk 12 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py]: I have changed the name to '_ENABLE_ALWAYS' with the leading underscore to further indicate that it is not a valid section name. I hope that's fine.

File: low-level/files/opt/seagate/sspl/bin/conf_sensors_enable (hunk 10 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/bin/conf_sensors_enable]: Done.

File: low-level/files/opt/seagate/sspl/bin/conf_sensors_enable (hunk 9 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/bin/conf_sensors_enable]: Done

File: low-level/framework/base/sspl_constants.py (hunk 11 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py:60]: These names correspond to pre-existing sections in the conf file.
The new names will also have to be sections.
AIUI these will have only the 'monitor' key.
Is that fine?

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 [hunk 15]
```diff
@@ -101,6 +101,7 @@ IP_addr=172.16.1.222
 max_login_attempts=2
 
 [RAIDSENSOR]
+monitor=true # RAIDSENSOR
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: why do we need these comments (like # RAIDSENSOR here) all over in config, does code depend on it. Had raised same earlier

File: low-level/framework/base/sspl_constants.py (hunk 11 above)
Comment (cdeshmukh) [reply low-level/framework/base/sspl_constants.py:60]: > These names correspond to pre-existing sections in the conf file.

What removed in this patch(in red), yes they are from existing config, i understand that.

> The new names will also have to be sections. AIUI these will have only the 'monitor' key. Is that fine?

I dont think we need so many changes to this patch further, just rename 'REALSTORSENSORS' to 'REALSTORMONITOR' on line 60. Same applies to other sections

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 15 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: The value of the 'monitor' key is changed using a shell script, specifically using 'sed'.
That works most naturally on single lines.

The comments help to enable/disable the value for the right sections. It should be possible to pick up the section name from the previous line, but that would be a bit more involved.

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        monitor[CPUFAULTSENSOR]="false"
+        ;;
+esac
+
+for group in "${!monitor[@]}"; do
+    sed_expr+=(-e 's/^monitor=true # '"${group}"'$/monitor='"${monitor[${group}]}"'/')
+done
+
+
+[ -f "${SSPL_CONF}" ] && {
```
[the remark this answers is not in the document]
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:47]: This file is executed from `low-level/files/opt/seagate/sspl/bin/sspl_post_install`, which also sources `./low-level/files/opt/seagate/sspl/bin/constants.sh`.
`SSPL_CONF` is defined there.

File: low-level/framework/base/sspl_constants.py (hunk 11 above)
Comment (pranav-seagate) [reply low-level/framework/base/sspl_constants.py:60]: > These names correspond to pre-existing sections in the conf file.
All the new top-level key names (`REALSTORSENSORS`, `REALSTORSENSORS`, etc.) correspond to sections in the config.
They must be changed in-sync with the names of sections themselves.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 15 above)
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:104]: >The comments help to enable/disable the value for the right sections. It should be possible to pick up the section name from the previous lines, but that would make the sed command somewhat more complicated.

thats needed, or if someone changes the comment, code will break, which is not right. So pick sections first & then run sed for respective monitor line.

File: low-level/framework/base/sspl_constants.py (hunk 11 above)
Comment (cdeshmukh) [reply low-level/framework/base/sspl_constants.py:60]: Ok so REALSTORMONITOR will have to section names you mean to address this comment. In that case, no we wont change that this time. Ultimately we need to later though, can you create a task tkt to address in LDR R2

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        monitor[SASPORTSENSOR]="false"
+        monitor[MEMFAULTSENSOR]="false"
+        monitor[CPUFAULTSENSOR]="false"
+        ;;
+esac
+
+for group in "${!monitor[@]}"; do
+    # We remove comments after the value since those are improperly parsed by the ConfigReader
+    # in sspl_test.
+    sed_expr+=(-e '/^monitor=.*$/{ x ; /^\['"${group}"'\]$/{ x ; s/^monitor=.*$/monitor='"${monitor[${group}]}"'/;  b} ; x }')
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:45]: btw config reader class also has interface to write back in same config file. So if you see this code creating problem, going pythonic way is definitely an option to consider and using those config reader interfaces.
check https://github.com/Seagate/cortx-monitor/pull/240/files#
salt_util.py -> def update_config_file() line 163

**Review by cdeshmukh — APPROVED** [review]

**Review by thavanathan — APPROVED** [review]

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable (hunk 17 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:47]: The sed command seems to be working so far.
But yes, if this get more complicated in the future we should consider moving to python, though that will have its own issues ( esp. calling into shell scripts from python and vice-versa during init. It may be required to change the entire file to python).

Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:47]: We can one have one simple python script which would just update required attribute in given section, with section/attribute names as cmd line arguments to the script. that way shell <-> python switch wont be felt that much

**Review by thavanathan — APPROVED** [review]

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+case "${storage_type}" in
+    virtual) 
+        monitor[REALSTORSENSORS]="false"
+        monitor[SYSTEMDWATCHDOG]="false"
+        ;;
+    jbod)
+        monitor[REALSTORSENSORS]="false"
+        monitor[SYSTEMDWATCHDOG]="false"
+        monitor[RAIDSENSOR]="false"
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:31]: SYSTEMDWATCHDOG & RAID sensor are for server disks & RAID arrays made of server disks again, why being disabled for external JBOD storage type ?

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    [SASPORTSENSOR]="true"
+    [MEMFAULTSENSOR]="true"
+    [CPUFAULTSENSOR]="true"
+
+)
+
+case "${storage_type}" in
+    virtual) 
+        monitor[REALSTORSENSORS]="false"
+        monitor[SYSTEMDWATCHDOG]="false"
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:26]: why systemdwatchdog sensor, meant for server local disk monitoring, disabled for external virtual storage type?

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        ;;
+    jbod)
+        monitor[REALSTORSENSORS]="false"
+        monitor[SYSTEMDWATCHDOG]="false"
+        monitor[RAIDSENSOR]="false"
+        ;;
+esac
+
+case "${server_type}" in
+    virtual)
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:36]: @ujjwalpl If server type is virtual, RAID sensor would needed still? meant for server local disk RAID monitoring?

File: sspl_test/framework/utils/config_reader.py [hunk 21]
```diff
@@ -44,7 +44,7 @@ def __init__(self):
             store_type = os.getenv('SSPL_STORE_TYPE', SSPL_STORE_TYPE)
             if store_type == StoreTypes.FILE.value:
                 self.store = configparser.RawConfigParser()
-                self.store.read(file_store_config_path)
+                self.store.read([file_store_config_path])
```
Comment (cdeshmukh) [inline sspl_test/framework/utils/config_reader.py:47]: why this change?

File: low-level/framework/base/sspl_constants.py (hunk 12 above)
Comment (cdeshmukh) [reply low-level/framework/base/sspl_constants.py]: yes

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable (hunk 18 above)
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:31]: i can recall this, so guess we are fine

> During JBOD discussions, which would be the case for most open src users, call was taken to keep minimal set of sensors enabled. Wherever open src users would install Cortx, may not have RAIDed disks, or have very less possibility and so the sensor we would disable by default.

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable (hunk 20 above)
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:36]: RAID may not be there in virtual env necessarily, safe is to keep it disabled by default. These configs can overridden if someone wants to try these sensors still in virtual env, like by editing its value in config file & not running this conf_based_sensors_enable

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable (hunk 17 above)
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:47]: We will go ahead with SED approach as @pranav-seagate is confident its working fine.

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable (hunk 19 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable]: Fixed as discussed here: https://teams.microsoft.com/_#/conversations/19:f13ac291ae4944db814d353d9c3f0b97@thread.v2?ctx=chat

File: low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable (hunk 18 above)
Comment (pranav-seagate) [reply low-level/files/opt/seagate/sspl/bin/conf_based_sensors_enable:30]: SystemdWatchdog should be enabled for JBOD.
RAID should be disabled for 'server_type' == 'virtual'. It should not depend on 'storage_type' at all.
As discussed.

**Review by cdeshmukh — APPROVED** [review]

**Review by thavanathan — APPROVED** [review]
